File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        for index, val in enumerate(nums):
            # index > 0 so we don't get an out of bound error with nums[index - 1]
            if index > 0 and val == nums[index - 1]:
                continue

            left, right = index + 1, len(nums) - 1

            while left < right:
                three_sum = val + nums[left] + nums[right]

                if three_sum > 0:
                    right -= 1
                elif three_sum < 0:
                    left +=1
                else:
                    res.append([val, nums[left], nums[right]])
                    left += 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1

        return res
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {

        // Initialize prefix to one which will accumulate product to the left of element
        int prefix = 1;

        // Intialize the postfix to one which will accumulate product to the right of the element
        int postfix = 1;

        // Intialize result array to be the same length of nums
        int[] res = new int[nums.Length];  
        
        /*
            Loop from left to right of the array
            The logic in the loop will make sure the prefix includes the product of all the elements to the left up to to current index
        */
        for(int i = 0; i < nums.Length; i++){

            // At the current index in the result array it will equal the prefix
            res[i] = prefix;
            
            // Then multiply the current prefix by the number at the current position in the array
            prefix *= nums[i];
        }
        
        /*
            Make a second pass through the array going backwards
            The logic in the loop will make sure the postfix includes the product of all elements to the right up to the current index
        */
        for(int i = nums.Length - 1; i >= 0; i--){

            // At the current index in the result it will equal postfix
            res[i] *= postfix;

            // Multiply the postfix by the number at the current position in the array
            postfix *= nums[i];
        }

        // Return the result array
        return res;
    }
}

